clients/pkg/logentry/stages/eventlogmessage.go

```diff
@@ -0,0 +1,164 @@
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+)
+
+const (
+	defaultSource                = "message"
+	ErrEmptyEvtLogMsgStageConfig = "empty event log message stage configuration"
+)
+
+type EventLogMessageConfig struct {
+	Source            *string `mapstructure:"source"`
+	DropInvalidLabels bool    `mapstructure:"drop_invalid_labels"`
+	OverwriteExisting bool    `mapstructure:"overwrite_existing"`
+}
+
+type eventLogMessageStage struct {
+	cfg    *EventLogMessageConfig
+	logger log.Logger
+}
+
+// Create a event log message stage, including validating any supplied configuration
+func newEventLogMessageStage(logger log.Logger, config interface{}) (Stage, error) {
+	cfg, err := parseEventLogMessageConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	// validate config (i.e., check that source is a valid log label)
+	err = validateEventLogMessageConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+	return &eventLogMessageStage{
+		cfg:    cfg,
+		logger: log.With(logger, "component", "stage", "type", "event_log_message"),
+	}, nil
+}
+
+// Parse the event log message configuration, creating a default configuration struct otherwise
+func parseEventLogMessageConfig(config interface{}) (*EventLogMessageConfig, error) {
+	cfg := &EventLogMessageConfig{}
+	err := mapstructure.Decode(config, cfg)
+	if err != nil {
+		return nil, err
+	}
+	return cfg, nil
+}
+
+// Ensure a event log message configuration object is valid, checking that any specified source
+// is a valid label name, and setting default values if a nil config object is provided
+func validateEventLogMessageConfig(c *EventLogMessageConfig) error {
+	if c == nil {
+		return errors.New(ErrEmptyEvtLogMsgStageConfig)
+	}
+	if c.Source != nil && !model.LabelName(*c.Source).IsValid() {
+		return fmt.Errorf(ErrInvalidLabelName, *c.Source)
+	}
+	return nil
+}
+
+func (m *eventLogMessageStage) Run(in chan Entry) chan Entry {
+	out := make(chan Entry)
+	key := defaultSource
+	if m.cfg.Source != nil {
+		key = *m.cfg.Source
+	}
+	go func() {
+		defer close(out)
+		for e := range in {
+			err := m.processEntry(e.Extracted, key)
+			if err != nil {
+				continue
+			}
+			out <- e
+		}
+	}()
+	return out
+}
+
+// Process a event log message from extracted with the specified key, adding additional
+// entries into the extracted map
+func (m *eventLogMessageStage) processEntry(extracted map[string]interface{}, key string) error {
+	value, ok := extracted[key]
+	if !ok {
+		if Debug {
+			level.Debug(m.logger).Log("msg", "source not in the extracted values", "source", key)
+		}
+		return nil
+	}
+	s, err := getString(value)
+	if err != nil {
+		if Debug {
+			level.Debug(m.logger).Log("msg", "invalid label value parsed", "value", value)
+		}
+		return err
+	}
+	lines := strings.Split(s, "\r\n")
+	for _, line := range lines {
+		parts := regexp.MustCompile(": ?").Split(line, 2)
```

Comment:
For efficiency the regex soups be compiled only once. I'm wondering if regex is required at all though.

Reply:
My alternative approach was going to be splitting on `:` and using strings.Trim* to remove the leading space(s) (if applicable), e.g.:

```go
parts := strings.SplitN(line, ":", 2)
...
mval := strings.TrimLeft(parts[1], " ")
```

Alternatively, it could just take off the first character if the value is not empty:

```go
mval := parts[1]
if len(mval) > 0 {
    mval = mval[1:]
}
```

Any preferences? If pushed, I guess I'd say I prefer the latter as it preserves the value as much as possible.

Reply:
Could you add a small benchmark and move the regex compile out? I think this is fine. We could follow up with a different version if it's faster.

Reply:
I've [added (hopefully fair!) benchmarks](https://github.com/grafana/loki/blob/88d2c29d6ffa48d491fffb9a9499c909ac1ae332/clients/pkg/logentry/stages/eventlogmessage_test.go#L344) for all three options:

```
goos: linux
goarch: amd64
pkg: github.com/grafana/loki/clients/pkg/logentry/stages
cpu: 12th Gen Intel(R) Core(TM) i7-12700H
BenchmarkSplittingKeyValuesRegex-20          	 1000000	      1158 ns/op	     614 B/op	       8 allocs/op
BenchmarkSplittingKeyValuesSplitTrim-20      	 1823691	       663.4 ns/op	     339 B/op	       5 allocs/op
BenchmarkSplittingKeyValuesSplitSubstr-20    	 2126520	       562.4 ns/op	     339 B/op	       5 allocs/op
```

```
goos: windows (VM)
goarch: amd64
pkg: github.com/grafana/loki/clients/pkg/logentry/stages
cpu: 12th Gen Intel(R) Core(TM) i7-12700H
BenchmarkSplittingKeyValuesRegex-4         	 1850302	       654.9 ns/op	     612 B/op	       8 allocs/op
BenchmarkSplittingKeyValuesSplitTrim-4     	 3055104	       389.8 ns/op	     338 B/op	       5 allocs/op
BenchmarkSplittingKeyValuesSplitSubstr-4   	 2888671	       374.5 ns/op	     338 B/op	       5 allocs/op
```

Looks to me like the two Split-based options are basically neck and neck - and given the string slicing option appears to be slightly more "correct", I'm going to go with it.